distribution_strategy/parameter_server_training/resnet_cifar_ps_strategy.py

```diff
@@ -0,0 +1,297 @@
+  with strategy.scope():
+    model = resnet_cifar_model.resnet56()
+
+    initial_learning_rate = (
+        img_class_common.BASE_LEARNING_RATE * BATCH_SIZE / 128)
+    # Using the learning rate schedule from the model garden:
+    # tensorflow_models/official/benchmark/models/resnet_cifar_main.py
+    lr_segments = [  # (multiplier, epoch to start) tuples
+        (0.1, 91), (0.01, 136), (0.001, 182)
+    ]
+    lr_schedule = tf.keras.optimizers.schedules.PiecewiseConstantDecay(
+        boundaries=list(p[1] * STEPS_PER_EPOCH for p in lr_segments),
+        values=[initial_learning_rate] +
+        list(p[0] * initial_learning_rate for p in lr_segments))
+    optimizer = img_class_common.get_optimizer(lr_schedule)
+
+    train_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
+        name="train_accuracy")
+    eval_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
+        name="eval_accuracy")
+
+    @tf.function
+    def worker_train_fn(iterator):
+
+      def replica_fn(inputs):
+        """Training loop function."""
+        batch_data, labels = inputs
+        with tf.GradientTape() as tape:
+          predictions = model(batch_data, training=True)
+          xent_loss = tf.keras.losses.SparseCategoricalCrossentropy(
+              reduction=tf.keras.losses.Reduction.NONE)(labels, predictions)
+          loss = (
+              tf.nn.compute_average_loss(xent_loss) +
+              tf.nn.scale_regularization_loss(model.losses))
+        gradients = tape.gradient(loss, model.trainable_variables)
+
+        optimizer.apply_gradients(
+            zip(gradients, model.trainable_variables))
+        train_accuracy.update_state(labels, predictions)
+        return loss
+
+      inputs = next(iterator)
+      losses = strategy.run(replica_fn, args=(inputs,))
+      return strategy.reduce(tf.distribute.ReduceOp.SUM, losses, axis=None)
+
+
+    @tf.function
+    def worker_eval_fn(iterator):
+
+      def eval_fn(inputs):
+        """Evaluation function"""
+        batch_data, labels = inputs
+        predictions = model(batch_data, training=False)
+        eval_accuracy.update_state(labels, predictions)
+
+      inputs = next(iterator)
+      strategy.run(eval_fn, args=(inputs,))
+
+    checkpoint_manager = tf.train.CheckpointManager(
+        tf.train.Checkpoint(model=model, optimizer=optimizer),
+        FLAGS.checkpoint_dir,
+        max_to_keep=2)
+    if checkpoint_manager.latest_checkpoint:
+      checkpoint = checkpoint_manager.checkpoint
+      checkpoint.restore(
+          checkpoint_manager.latest_checkpoint
+          ).assert_existing_objects_matched()
+
+    train_dataset_fn = lambda _: cifar_preprocessing.input_fn(
+        is_training=True,
+        data_dir=FLAGS.data_dir,
+        batch_size=BATCH_SIZE,
+        parse_record_fn=cifar_preprocessing.parse_record,
+        dtype=tf.float32,
+        drop_remainder=True)
+    eval_dataset_fn = lambda _: cifar_preprocessing.input_fn(
```

Comment:
Is the eval data shuffled? If not, could you add a comment and a TODO?

Reply:
Maybe you can just append a shuffle at the end of the dataset?

Reply:
input_fn already shuffles the training data using process_record_dataset: [code link](https://github.com/tensorflow/models/blob/871c4e0a393ef4385534bee55354a5df8aa1ccf4/official/legacy/image_classification/resnet/imagenet_preprocessing.py#L108)